columnar/src/columnar/merge/mod.rs

```diff
@@ -348,56 +358,157 @@ fn is_empty_after_merge(
     }
 }
 
-#[allow(clippy::type_complexity)]
-fn group_columns_for_merge(
-    columnar_readers: &[&ColumnarReader],
-    required_columns: &[(String, ColumnType)],
-    merge_row_order: &MergeRowOrder,
-) -> io::Result<BTreeMap<(String, ColumnType), Vec<Option<DynamicColumn>>>> {
-    // Each column name may have multiple types of column associated.
-    // For merging we are interested in the same column type category since they can be merged.
-    let mut columns_grouped: HashMap<(String, ColumnTypeCategory), GroupedColumns> = HashMap::new();
-
-    for &(ref column_name, column_type) in required_columns {
-        columns_grouped
-            .entry((column_name.clone(), column_type.into()))
-            .or_insert_with(|| {
-                GroupedColumns::for_category(column_type.into(), columnar_readers.len())
-            })
-            .require_type(column_type)?;
+type MergeIter<'a> =
+    Box<dyn Iterator<Item = io::Result<(Rc<str>, ColumnType, GroupedColumns)>> + 'a>;
+
+#[derive(Debug, Clone)]
+struct MergeColumn {
+    column_name: Rc<str>,
+    reader_ord: usize,
+    column: DynamicColumnHandle,
+}
+impl MergeColumn {
+    fn new(column_name: Rc<str>, reader_ord: usize, column: DynamicColumnHandle) -> Self {
+        MergeColumn {
+            column_name,
+            reader_ord,
+            column,
+        }
     }
+}
 
-    for (columnar_id, columnar_reader) in columnar_readers.iter().enumerate() {
-        let column_name_and_handle = columnar_reader.list_columns()?;
-        // We skip columns that end up with 0 documents.
-        // That way, we make sure they don't end up influencing the merge type or
-        // creating empty columns.
-
-        for (column_name, handle) in column_name_and_handle {
-            let column_category: ColumnTypeCategory = handle.column_type().into();
-            let column = handle.open()?;
-            if is_empty_after_merge(merge_row_order, &column, columnar_id) {
-                continue;
+/// Iterates over the columns of the columnar readers, grouped by column name.
+/// Key functionality is that `open` of the Columns is done lazy per group.
+fn group_columns_for_merge_iter<'a>(
```

Comment:
The code is much more complicated. Could we have kept the previous code, and worked of `DynamicColumnHandle` instead of `DynamicColumn` and gotten the same result?

Reply:
The columns need to be loaded to access things like min max values. The alternative would be to change the code so we can access only the metadata with something like `open_metadata`

Reply:
We can lighten the work of `group_columns_for_merge` and change it so that it does not:
- filter out columns that end up being empty.
- identify the resulting ColumnType.

It would just stops at returning a 

```
fn group_columns_for_merge(
    columnar_readers: &[&ColumnarReader],
    required_columns: &[(String, ColumnType)],
    merge_row_order: &MergeRowOrder,
) -> io::Result<BTreeMap<(String, ColumnTypeCategory), Vec<Option<DynamicColumnHandle>>>> {
```

On the callsite, we would then iterate through these handles, and open the `DynamicColumn`,
and do the conversion of the `(ColumnTypeCategory, Vec<Option<DynamicColumnHandle>>)` into a 
`(ColumnType, Vec<Option<DynamicHandle>>)` where all of the columns that are Some contain at least one element.

The rest is then similar to what exists today.
The code should not be more complicated than it is today.

One tricky part to consider is how to deal with required columns. 
If all columns for this `(String, ColumnTypeCategory)` are empty (following deletes), whether or not we want a column depends on whether the column is marked as required or not.

Reply:
In the high cardinality num column case, the BTreeMap variant may cost more than 100MB heap allocation, which is why I opted to the Iterator version.
From the code complexity the main difference is the mixing of the required columns into the (sorted by column name) Column stream instead of simply inserting them into the `BTreeMap`

Reply:
This is too complex. Can you go with the BTreeMap solution?

Reply:
I switched to a BtreeMap